### PR TITLE
BinaryenSetFunctionTable now accepts array of func names not funcs.

### LINKED ISSUE #1645: BinaryenSetFunctionTable cannot (safely) accept imported functions
In wasm you're allowed to indirectly call imported functions, so their index must be in the function table. However Binaryen's C-API `BinaryenSetFunctionTable` currently only safely supports passing an array of `BinaryenFunctionRef`, which are defined with `BinaryenAddFunction()`. Function imports are `BinaryenImportRef`s.

If you're using C++ you could reach behind the curtain and add the table segments yourself, however when this is isn't possible you're only other option is to unsafely cast `BinaryenImportRef` to `BinaryenFunctionRef` which _actually works_ because both `wasm::Function` and `wasm::Import` happen to have `std::string name` as their first field, but it might not always be that way so this is pretty unsafe.

```c
BinaryenModuleRef module = BinaryenModuleCreate();
BinaryenFunctionTypeRef fun_type = BinaryenAddFunctionType(module, NULL, BinaryenTypeNone(), NULL, 0);
BinaryenImportRef fun_import = BinaryenAddFunctionImport(module, "example", "env", "example", fun_type);

// YOLO, this works because wasm::Function and wasm::Import both have std::string name as
// their first field and that's all that matters, but it might not always be that way
BinaryenFunctionRef fun = (BinaryenFunctionRef)fun_import;

BinaryenFunctionRef fun_table[] = { fun };
BinaryenSetFunctionTable(module, fun_table, 1);
```

Since the name of the function is what is actually used, perhaps `BinaryenSetFunctionTable` could be changed to accept `const char **` instead?

***

btw--for my future reference how does binaryen handle breaking changes? haha :trollface:

### PR DESCRIPTION
**Pull request by jayphelps**
fixes #1645.

I included a changelog file per request.

### CODE REVIEW

File: src/binaryen-c.cpp [hunk 1]
```diff
@@ -1882,28 +1882,27 @@ void BinaryenRemoveExport(BinaryenModuleRef module, const char* externalName) {
 
 // Function table. One per module
 
-void BinaryenSetFunctionTable(BinaryenModuleRef module, BinaryenFunctionRef* funcs, BinaryenIndex numFuncs) {
+void BinaryenSetFunctionTable(BinaryenModuleRef module, const char **funcNames, BinaryenIndex numFuncNames) {
   if (tracing) {
     std::cout << "  {\n";
-    std::cout << "    BinaryenFunctionRef funcs[] = { ";
-    for (BinaryenIndex i = 0; i < numFuncs; i++) {
+    std::cout << "    const char* funcNames[] = { ";
```
Comment (jayphelps) [inline src/binaryen-c.cpp]: btw `char* funcNames` vs ``char *funcNames` pointer placement isn't consistent in the codebase but I tried to be consistent with other similar contexts. Happy to follow any pattern.

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+significant internal modifications and optimizations etc. generally deserve a
+mention. To examine the full set of changes between versions, visit the link to
+full changeset diff at the end of each section.
+
+Current Trunk
+-------------
+
+### BREAKING CHANGES
+
+- `BinaryenSetFunctionTable` no longer accepts an array of functions, instead it accepts an array of function names, `const char **funcNames`. Previously, you could not include imported functions because they are of type `BinaryenImportRef` instead of `BinaryenFunctionRef`. [#1650](https://github.com/WebAssembly/binaryen/pull/1650)
```
Comment (kripken) [inline CHANGELOG.md]: please mention that this is in the C API

File: src/binaryen-c.cpp (hunk 1 above)
Comment (kripken) [reply src/binaryen-c.cpp]: Yeah, we should be more consistent. I think we prefer `char* funcNames` (keep the type all together).

File: src/js/binaryen.js-post.js [hunk 3]
```diff
@@ -1129,9 +1129,17 @@ Module['Module'] = function(module) {
...
     });
   };
-  this['setFunctionTable'] = function(funcs) {
+  this['setFunctionTable'] = function(funcNames) {
     return preserveStack(function() {
-      return Module['_BinaryenSetFunctionTable'](module, i32sToStack(funcs), funcs.length);
+      return Module['_BinaryenSetFunctionTable'](
+        module,
+        i32sToStack(
+          funcNames.map(function(funcName) {
```
Comment (kripken) [inline src/js/binaryen.js-post.js]: can use `strToStack` here (see examples above in the same file)

**Comment by kripken** [comment]
Looks like the test failure is in the JS API code - the kitchen sink test fails on `[wasm-validator error in module] unexpected false: segment name should be valid`, so perhaps the name isn't being passed through properly.

**Comment by jayphelps** [comment]
Updated. I had needed to rebuild the js files. 4a6a2f7ebdb63c37800458ab55ecc5404a15c74b feel free to squash after review.

File: src/js/binaryen.js-post.js [hunk 4]
```diff
@@ -1532,6 +1526,11 @@ Module['getFunctionInfo'] = function(func) {
   };
 };
 
+// Obtains name for a 'Function'
+Module['getFunctionName'] = function(func) {
```
[the remark this answers is not in the document]
Comment (jayphelps) [reply src/js/binaryen.js-post.js]: Added for parity with the other kitchen sink demo but also cause the general info call defers to a number of other calls and I figured you’d recommend exposing this one lol. No problem! Changing 🤙

**Comment by kripken** [comment]
Great, thanks!

Btw, I'm not opposed to adding new functions like that - just I was surprised it was here, I thought I'd missed something. Can discuss that in a separate issue if you want.

**Pull request merged**